snapcraft/internal/sources/_base.py

```diff
@@ -45,12 +46,43 @@ def __init__(self, source, source_dir, source_tag=None, source_commit=None,
 class FileBase(Base):
 
     def pull(self):
-        if snapcraft.internal.common.isurl(self.source):
-            self.download()
-        else:
-            shutil.copy2(self.source, self.source_dir)
+        file_cache = FileCache()
+        source_file = None
+        is_source_url = snapcraft.internal.common.isurl(self.source)
+
+        # First check if we already have the source file cached.
+        if self.source_checksum:
+            algorithm, hash = split_checksum(self.source_checksum)
+            cache_file = file_cache.get(algorithm=algorithm, hash=hash)
+            if cache_file:
+                source_file = os.path.join(self.source_dir,
+                                           os.path.basename(cache_file))
+                # We make this copy as the provisioning logic can delete
+                # this file and we don't want that.
+                shutil.copy2(cache_file, source_file)
```

Comment:
I feel this might be cleaner if file_cache.get could just take source_checksum even if it's None. Though you can probably argue that more explicit is clearer, so feel free to leave it.